File: ios/Sqwack/DesignSystem/Components.swift
import SwiftUI

// Shared visual components for the mockup design language.

/// Tiny line chart. Values are auto-normalized; flat/empty data draws a baseline.
struct Sparkline: View {
    let values: [Double]
    var color: Color = .blue

    var body: some View {
        GeometryReader { geo in
            let maxValue = max(values.max() ?? 1, 0.001)
            let points = values.isEmpty ? [0, 0] : values
            let stepX = geo.size.width / CGFloat(max(points.count - 1, 1))
            Path { path in
                for (i, value) in points.enumerated() {
                    let x = CGFloat(i) * stepX
                    let y = geo.size.height * (1 - CGFloat(value / maxValue) * 0.9) - 1
                    if i == 0 { path.move(to: CGPoint(x: x, y: y)) } else { path.addLine(to: CGPoint(x: x, y: y)) }
                }
            }
            .stroke(color, style: StrokeStyle(lineWidth: 1.5, lineCap: .round, lineJoin: .round))
        }
    }
}

/// Provider avatar: official icon when available, fallback glyph otherwise.
struct ProviderBadge: View {
    let provider: String
    var size: CGFloat = 44

    private var normalizedProvider: String {
        provider.lowercased()
    }

    private var assetName: String? {
        switch normalizedProvider {
        case "claude": "ProviderClaude"
        case "codex": "ProviderOpenAI"
        case "deepseek": "ProviderDeepSeek"
        case "openai", "openai api": "ProviderOpenAI"
        default: nil
        }
    }

    private var color: Color {
        switch normalizedProvider {
        case "claude": .orange
        case "codex": .blue
        case "openai", "openai api": .green
        case "deepseek": .purple
        default: .gray
        }
    }

    private var symbol: String {
        switch normalizedProvider {
        case "claude": "rays"
        case "codex": "cube.fill"
        case "openai", "openai api": "sparkles"
        default: "sparkles"
        }
    }

    var body: some View {
        ZStack {
            Circle().fill(Color.consolePanelRaised)
            if let assetName {
                Image(assetName)
                    .resizable()
                    .scaledToFit()
                    .padding(size * 0.18)
            } else {
                Circle().fill(color.gradient)
                Image(systemName: symbol)
                    .font(.system(size: size * 0.45, weight: .bold))
                    .foregroundStyle(.white)
            }
        }
        .overlay(Circle().strokeBorder(Color.consoleStrokeBright))
        .frame(width: size, height: size)
    }
}

/// Rounded panel container used across Overview and Development.
struct Panel<Content: View>: View {
    var title: String?
    var badge: String?
    var trailing: String?
    @ViewBuilder var content: Content

    var body: some View {
        VStack(alignment: .leading, spacing: 14) {
            if title != nil || badge != nil {
                HStack(spacing: 10) {
                    if let title {
                        Text(title)
                            .font(.headline)
                    }
                    if let badge {
                        Text(badge)
                            .font(.subheadline.weight(.semibold))
                            .padding(.horizontal, 8)
                            .padding(.vertical, 2)
                            .background(Capsule().fill(Color(.tertiarySystemFill)))
                            .foregroundStyle(.secondary)
                    }
                    Spacer()
                    if let trailing {
                        Text(trailing)
                            .font(.subheadline)
                            .foregroundStyle(.blue)
                    }
                }
            }
            content
        }
        .padding(20)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(
            RoundedRectangle(cornerRadius: 16)
                .fill(Color.consolePanel)
                .overlay(RoundedRectangle(cornerRadius: 16).strokeBorder(Color.consoleStroke))
        )
    }
}

/// Small filled progress bar for resource meters.
struct MeterBar: View {
    let fraction: Double
    var color: Color = .blue
    var height: CGFloat = 6

    var body: some View {
        GeometryReader { geo in
            ZStack(alignment: .leading) {
                Capsule().fill(Color.white.opacity(0.08))
                Capsule()
                    .fill(color)
                    .frame(width: max(height, geo.size.width * min(max(fraction, 0), 1)))
            }
        }
        .frame(height: height)
    }
}

/// Capsule chip with an SF Symbol, as under agent rows in the mockup.
struct Chip: View {
    let icon: String
    let text: String

    var body: some View {
        HStack(spacing: 5) {
            Image(systemName: icon).font(.caption2)
            Text(text).font(.caption)
        }
        .padding(.horizontal, 9)
        .padding(.vertical, 4)
        .background(Capsule().fill(Color.white.opacity(0.07)))
        .foregroundStyle(.secondary)
    }
}

enum Format {
    static func bytes(_ value: Int64) -> String {
        let gb = Double(value) / 1_073_741_824
        if gb >= 1 { return String(format: gb >= 10 ? "%.0f GB" : "%.1f GB", gb) }
        return String(format: "%.0f MB", Double(value) / 1_048_576)
    }

    static func uptime(_ seconds: Int) -> String {
        let days = seconds / 86400
        let hours = seconds % 86400 / 3600
        if days > 0 { return "\(days)d \(hours)h" }
        let minutes = seconds % 3600 / 60
        return hours > 0 ? "\(hours)h \(minutes)m" : "\(minutes)m"
    }
}

extension AgentState {
    /// Sentence-case label for list rows ("Working"), vs. the big card label.
    var shortLabel: String { label.capitalized }
}

File: android/app/src/main/java/com/sqwack/Components.kt
package com.sqwack

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.clickable
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.foundation.Image
import java.util.Locale
import kotlin.math.max

// Shared visual components for the mockup design language.

/// Tiny line chart. Values are auto-normalized; flat/empty data draws a baseline.
@Composable
fun Sparkline(values: List<Double>, color: Color = Palette.blue, modifier: Modifier = Modifier) {
    Canvas(modifier.fillMaxSize()) {
        val points = if (values.isEmpty()) listOf(0.0, 0.0) else values
        val maxValue = max(points.max(), 0.001)
        val stepX = size.width / max(points.size - 1, 1)
        val path = Path()
        points.forEachIndexed { index, value ->
            val x = index * stepX
            val y = size.height * (1f - (value / maxValue).toFloat() * 0.9f) - 1f
            if (index == 0) path.moveTo(x, y) else path.lineTo(x, y)
        }
        drawPath(path, color, style = Stroke(width = 1.5.dp.toPx(), cap = StrokeCap.Round, join = StrokeJoin.Round))
    }
}

/// Provider avatar: official icon when available, fallback glyph otherwise.
@Composable
fun ProviderBadge(provider: String, size: Dp = 44.dp) {
    val normalized = provider.lowercase(Locale.US)
    val asset = when (normalized) {
        "claude" -> R.drawable.provider_claude
        "codex", "openai", "openai api" -> R.drawable.provider_openai
        "deepseek" -> R.drawable.provider_deepseek
        else -> null
    }
    val fallbackColor = when (normalized) {
        "hermes" -> Palette.mint
        else -> Palette.gray
    }
    Box(
        Modifier
            .size(size)
            .clip(CircleShape)
            .background(Palette.panelRaised)
            .border(1.dp, Palette.strokeBright, CircleShape),
        contentAlignment = Alignment.Center,
    ) {
        if (asset != null) {
            Image(
                painterResource(asset),
                contentDescription = provider,
                contentScale = ContentScale.Fit,
                modifier = Modifier.padding(size * 0.18f).fillMaxSize(),
            )
        } else {
            Box(
                Modifier
                    .fillMaxSize()
                    .background(Brush.linearGradient(listOf(fallbackColor, fallbackColor.copy(alpha = 0.6f)))),
                contentAlignment = Alignment.Center,
            ) {
                Text(
                    provider.take(1).uppercase(Locale.US),
                    color = Color.White,
                    fontWeight = FontWeight.Bold,
                    fontSize = (size.value * 0.45f).sp,
                )
            }
        }
    }
}

/// Rounded panel container used across Overview and Development.
@Composable
fun Panel(
    modifier: Modifier = Modifier,
    title: String? = null,
    badge: String? = null,
    trailing: String? = null,
    content: @Composable androidx.compose.foundation.layout.ColumnScope.() -> Unit,
) {
    Column(
        modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(16.dp))
            .background(Palette.panel)
            .border(1.dp, Palette.stroke, RoundedCornerShape(16.dp))
            .padding(20.dp),
        verticalArrangement = Arrangement.spacedBy(14.dp),
    ) {
        if (title != null || badge != null) {
            Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(10.dp)) {
                if (title != null) Text(title, fontSize = 17.sp, fontWeight = FontWeight.Bold, color = Palette.primaryText)
                if (badge != null) {
                    Text(
                        badge,
                        fontSize = 13.sp,
                        fontWeight = FontWeight.SemiBold,
                        color = Palette.secondaryText,
                        modifier = Modifier
                            .clip(CircleShape)
                            .background(Palette.fill)
                            .padding(horizontal = 8.dp, vertical = 2.dp),
                    )
                }
                Spacer(Modifier.weight(1f))
                if (trailing != null) Text(trailing, fontSize = 13.sp, color = Palette.blue)
            }
        }
        content()
    }
}

/// Small filled progress bar for resource meters.
@Composable
fun MeterBar(fraction: Double, color: Color = Palette.blue, height: Dp = 6.dp, modifier: Modifier = Modifier) {
    Box(
        modifier
            .fillMaxWidth()
            .height(height)
            .clip(CircleShape)
            .background(Color.White.copy(alpha = 0.08f)),
    ) {
        Box(
            Modifier
                .fillMaxHeight()
                .fillMaxWidth(fraction.coerceIn(0.0, 1.0).toFloat())
                .clip(CircleShape)
                .background(color),
        )
    }
}

/// Capsule chip, as under agent rows in the mockup.
@Composable
fun Chip(icon: String, text: String) {
    Row(
        Modifier
            .clip(CircleShape)
            .background(Palette.fill)
            .padding(horizontal = 9.dp, vertical = 4.dp),
        horizontalArrangement = Arrangement.spacedBy(5.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(icon, fontSize = 10.sp, color = Palette.secondaryText)
        Text(text, fontSize = 12.sp, color = Palette.secondaryText, maxLines = 1)
    }
}

@Composable
fun Dot(color: Color, size: Dp = 8.dp, modifier: Modifier = Modifier) {
    Box(modifier.size(size).clip(CircleShape).background(color))
}

@Composable
fun Divider() = HorizontalDivider(color = Palette.stroke)

@Composable
fun Mono(text: String, fontSize: androidx.compose.ui.unit.TextUnit = 13.sp, color: Color = Palette.secondaryText, fontWeight: FontWeight = FontWeight.Normal, modifier: Modifier = Modifier) {
    Text(text, fontSize = fontSize, color = color, fontFamily = FontFamily.Monospace, fontWeight = fontWeight, maxLines = 1, modifier = modifier)
}

/// Pill used for tab buttons and menu triggers.
@Composable
fun Pill(
    selected: Boolean = false,
    onClick: () -> Unit,
    padding: PaddingValues = PaddingValues(horizontal = 14.dp, vertical = 8.dp),
    content: @Composable androidx.compose.foundation.layout.RowScope.() -> Unit,
) {
    Row(
        Modifier
            .clip(CircleShape)
            .background(if (selected) Palette.blue.copy(alpha = 0.14f) else Palette.panelRaised)
            .border(1.dp, Palette.stroke, CircleShape)
            .clickable(onClick = onClick)
            .padding(padding),
        horizontalArrangement = Arrangement.spacedBy(6.dp),
        verticalAlignment = Alignment.CenterVertically,
        content = content,
    )
}
